core/device_hooks/common_kernels.inc.cpp

```diff
@@ -391,6 +397,12 @@ GKO_NOT_COMPILED(GKO_HOOK_MODULE);
 GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(
     GKO_DECLARE_COO_CONVERT_ROW_IDXS_TO_PTRS_KERNEL);
 
+template <typename ValueType, typename IndexType>
+GKO_DECLARE_COO_CONVERT_TO_CSR_KERNEL(ValueType, IndexType)
+GKO_NOT_COMPILED(GKO_HOOK_MODULE);
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_COO_CONVERT_TO_CSR_KERNEL);
```

Comment:
Comment for @ginkgo-project/developers :
We (@fritzgoebel and I) decided to always have an explicit `XXX_to_YYY_KERNEL` to make things uniform.
Previously, there was no COO to CSR kernel. In the conversion, only
```c++
coo::convert_row_idxs_to_ptrs
```
was called, which means there was no need for an explicit conversion kernel. We decided to create a kernel 
```c++
coo::convert_to_csr(Csr *result, Coo *source)  // simplified for better readability
```
which basically only calls `coo::make_convert_row_idxs_to_ptrs`. There should not be any performance penalty (before our change, a function `coo::make_csr` inside `core/matrix/coo.cpp` was called, so the number of function calls are the same), we just moved the `convert_row_idxs_to_ptrs` call from `core` to the `convert_to_csr` kernel function (which is inside `reference`, `omp` and `cuda`).

Reply:
Personally, I don't get the benefits of uniformity here. It just limits the available ways to reduce code duplication and find the correct balance between common and platform-specific code, without having any obvious benefits (at least I don't see one). The only reason I can think of for doing it would be generating code that uses these kernels automatically (which I do not think is in the agenda).

If we already are going for this solution, I would at least suggest to name the kernels just `convert(Csr *result, Coo *source)`. In C++ parameter types are a part of the signature, so adding the `_to_csr` at the end is completely redundant (the same could be argued for the `coo::` namespace). `_to_csr` actually results in the same problems as the ones you encounter when using BLAS-like interfaces with their explicit naming of the type as part of the function name. If you try to write a template around it that is type-agnostic, using the `convert_XXX_to_YYY` convention would require writing a template specialization that maps all of these kernels to the same call, where there would be no need to do it when using just `convert` and relying on C++'s name mangling. Check out the [cuBLAS](https://github.com/ginkgo-project/ginkgo/blob/develop/cuda/base/cublas_bindings.hpp) and [cuSPARSE bindings](https://github.com/ginkgo-project/ginkgo/blob/develop/cuda/base/cusparse_bindings.hpp) in Ginkgo for a real-world demonstration of the problem.

Reply:
In the file `include/ginkgo/core/matrix/csr.hpp`, we currently have a protected member `std::unique_ptr<Coo<ValueType, IndexType>> make_coo()`, which basically does the exact same thing as the already existing function `void convert_to(Coo<ValueType, IndexType> *result)` (because the last one calls `make_coo`). I don't think it is a good idea to have an explicit function `make_coo()` when we already have the `convert_to` function.

We basically do the exact same thing in this PR with the conversion kernels that we do for all the other conversions, which is having an explicit conversion kernel.
The only benefit I can think of here is that it makes it a bit easier for developer to know how these conversions work because all work in the same manner.

I am also fine reverting it back to the previous state, however, I would prefer to remove the currently `protected` member functions `make_csr()` and `make_coo()` and just making them local functions inside `core/matrix/coo.cpp` and `core/matrix/csr.cpp`, so the user never sees them (in for example a UML class diagramm).

Reply:
This is a summary what we did (step 3 is still in the making) in regards to the removing / renaming of kernels.

We did the following changes for the CSR -> COO conversion (similarly for COO -> CSR):
1. removed the `make_coo` from the `Csr` class
2. added a kernel `convert_to_coo` which does the exact same thing as the `make_coo` did (calling `convert_row_ptrs_to_idxs` without anything else)
3. drop the kernel `convert_row_ptrs_to_idxs` from being publicly known (only `convert_to_coo` needs to be known)
4. the function `convert_to(Coo<...> *result)` calls `convert_to_coo` now instead of `make_coo`

I prefer having a `convert_to_YYY` kernel call in a `convert_to` function than a `convert_row_ptrs_to_idxs` kernel in one or two specific conversions, so we keep it uniform with all the other conversions and it is easier to understand what is done inside the conversion.